### PR TITLE
feat(encoder): strategy-aware literal gates (G4 + G5)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Donor-parity `ZSTD_minLiteralsToCompress` (G4) and `ZSTD_minGain` (G5)
wired into literal-section gates on both emit and estimator paths.
Companion to G3 (whole-block bail-out, #181).

### G4 — `ZSTD_minLiteralsToCompress` (donor `zstd_compress_literals.c:114-127`)

Strategy-aware floor `8 << shift`, 6-byte floor on reused HUF tables:

| strategy band | shift | floor (no HUF reuse) |
|---|---|---|
| fast / dfast / greedy / lazy | 3 | 64 |
| btopt | 2 | 32 |
| btultra | 1 | 16 |
| btultra2 | 0 | 8 |

Previously hardcoded `< 8` — same lower bound as btultra2 but ignored
strategy.

### G5 — `ZSTD_minGain` (donor `zstd_compress_internal.h:677-684`)

Strategy-aware expansion margin `(srcSize >> minlog) + 2`:

| strategy band | minlog | margin |
|---|---|---|
| fast..btopt | 6 | ~1.56% + 2 |
| btultra | 7 | ~0.78% + 2 |
| btultra2 | 8 | ~0.39% + 2 |

G5 comparison uses donor formula `cLitSize >= srcSize - minGain` form
(payload + tree_desc vs literal payload length, no headers on either
side). Previously bare `>= raw_section_bytes` (zero margin) and earlier
iteration of `>= raw_section_bytes - mg` (which skewed by header diff)
forced fallback on marginally-winning compressed sections that donor
would keep.

### Plumbing + estimator/emit parity

`CompressState.strategy_tag` threaded through `FrameCompressor::new` /
`new_with_matcher` / `reset` and synced into `CompressedBlockScratch`'s
`scratch_state`. Reset-time sync happens at both reset sites:
`FrameCompressor::compress` and `StreamingEncoder::ensure_frame_started`.
Centralized `use_raw_literal_fallback` helper shared by emit
(`compress_literals`) and probe (`estimate_literals_section_bytes`) —
single decision point, no drift.

Emit's RLE pre-shortcut fires for **any non-empty all-identical literal
section** (was gated by `len >= 8`). RLE and raw share the same lhSize
for a given `len`, so RLE = lhSize + 1 equals raw = lhSize + len at
`len == 1` and is smaller by `len - 1` bytes for `len >= 2`. The
pre-check now correctly covers:
- `len < min_lits` all-identical (donor falls through to raw — we win)
- `len >= min_lits` all-identical (donor reaches the same RLE block
  via `cLitSize == 1` after compress — same wire output)

The pre-check runs BEFORE the `min_lits` gate, and
`estimate_literals_section_bytes` mirrors this order so splitter probe
costs match real emit byte-for-byte.

## Verification

- 534/534 lib tests pass (debug profile), including 6 new tests:
  - `min_literals_to_compress_returns_per_strategy_floor` — per-strategy
    floor + HUF reuse
  - `min_gain_returns_per_strategy_margin` — per-strategy margin +
    small-source edge
  - `use_raw_literal_fallback_uses_payload_vs_srcsize_threshold` —
    walks the 2-byte gap between the donor payload-vs-srcSize formula
    and the on-wire-vs-on-wire form at `literals.len() = 20`
  - `estimator_literals_section_mirrors_emit_for_short_inputs` —
    byte-equivalent estimator/emit on boundary lengths around each
    strategy's `min_lits`, sub-`min_lits` all-identical (RLE
    pre-check), and the HUF-reuse path with seeded `last_huff_table`
  - `ensure_frame_started_refreshes_stale_strategy_tag_at_reset` —
    streaming reset-time sync: deliberately corrupts
    `state.strategy_tag` to a sentinel before the first write,
    asserts `ensure_frame_started` restores the correct tag
  - `set_compression_level_then_compress_refreshes_strategy_tag` —
    `FrameCompressor::set_compression_level` from `Fastest` to
    `Level(20)` (Fast band → BtUltra2 band) followed by `compress()`
    must produce the new band's resolved tag
- clippy release all-targets `-D warnings` — clean
- fmt check — clean
- Full ratio matrix (203 cells via REPORT lines): 0 deltas vs main
- Speed (decodecorpus-z000033, criterion 10 samples, clean back-to-back):
  - L1 fast / L2 dfast / L3 dfast / L16 btopt — no significant change
    vs main (all p > 0.4 in latest clean run)
  - L20/L21/L22 btultra2 high-entropy-1m — no change vs main (p > 0.05),
    Rust ~10× faster than FFI; fast-skip preserved
  - L8 lazy: −8.4% vs main *(pre-existing lazy regression; tracked
    separately as next-up lazy investigation, not introduced by this
    PR)*

## Notes / Follow-ups

- L8 lazy −8% — separate lazy track (ratio + speed) follows this PR.
- Block-level margin in `SplitEstimator::estimate_subblock_size` /
  `emit_single_sequence_block` still uses uniform `(source_len >> 8) + 2`
  (btultra2 value applied across all strategies). The `min_gain` helper
  is currently wired only into literal-section gates; migration to
  the block-level sites is a separate cleanup.
- Remaining block-splitting gaps: G6 (`HUF_flags_preferRepeat`), G1
  (superblock shared entropy), G2 (cost-budgeted carve), G7 (extend
  post-split to lazy band) — tracked in #23.

Related: #23

## Summary by CodeRabbit

* **Refactor**
  * Optimized internal compression literal-encoding logic for improved consistency and efficiency across compression levels.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/182?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/encoding/blocks/compressed.rs [hunk 1]
```diff
@@ -437,13 +524,21 @@ fn estimate_literals_section_bytes(
...
+    // Mirror emit's RLE gate exactly: emitter only emits RLE when
+    // `len >= 8 && all_identical` (see comment at the RLE branch above).
+    // With reused HUF tables `min_lits == 6`, so a bare `all_identical`
+    // check here would have estimator predict a 1-byte RLE payload for
+    // 6-7 byte all-identical literals that the emitter would actually
+    // write as raw — splitter probe costs drift from real emit output.
+    if literals.len() >= 8 && all_bytes_identical(literals) {
         *last_huff = None;
         return uncompressed_literals_header_bytes(literals.len()) + 1;
     }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/blocks/compressed.rs]: @coderabbitai Fixed in 42034fc7 — reordered `estimate_literals_section_bytes` so the RLE pre-check runs before `min_lits` (matching emit). Also added an estimator/emit parity test that covers boundary lengths around each strategy's `min_lits` and the RLE `>= 8` cliff.

**Comment by polaz** [comment]
@coderabbitai full review
